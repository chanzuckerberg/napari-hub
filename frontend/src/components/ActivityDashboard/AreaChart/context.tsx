import { createContext, ReactNode, useContext, useMemo } from 'react';

interface AreaChartStateValue {
  publicReleaseX: number;
  isScreen600: boolean;
  visibleMonths: number[];
}

const AreaChartStateContext = createContext<AreaChartStateValue>({
  publicReleaseX: 0,
  isScreen600: false,
  visibleMonths: [],
});

interface Props extends AreaChartStateValue {
  children: ReactNode;
}

/**
 * Provider for sharing state used in the area chart component.
 */
export function AreaChartStateProvider({
  children,
  isScreen600,
  publicReleaseX,
  visibleMonths,
}: Props) {
  const value = useMemo<AreaChartStateValue>(
    () => ({
      publicReleaseX,
      isScreen600,
      visibleMonths,
    }),
    [isScreen600, publicReleaseX, visibleMonths],
  );

  return (
    <AreaChartStateContext.Provider value={value}>
      {children}
    </AreaChartStateContext.Provider>
  );
}

export function useAreaChartState(): AreaChartStateValue {
  return useContext(AreaChartStateContext);
}
