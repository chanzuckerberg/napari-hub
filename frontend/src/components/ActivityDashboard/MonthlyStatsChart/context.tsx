import dayjs from 'dayjs';
import { createContext, ReactNode, useContext, useMemo } from 'react';

import { I18nKeys } from '@/types/i18n';

interface MonthlyStatsValue {
  dateLineI18nKey: I18nKeys<'activity'>;
  startDate: dayjs.ConfigType;
}

const MonthlyStatsContext = createContext<MonthlyStatsValue | null>(null);

interface Props
  extends Pick<MonthlyStatsValue, 'startDate' | 'dateLineI18nKey'> {
  children: ReactNode;
}

export function MonthlyStatsProvider({
  children,
  dateLineI18nKey,
  startDate,
}: Props) {
  const value = useMemo(
    () => ({
      dateLineI18nKey,
      startDate,
    }),
    [dateLineI18nKey, startDate],
  );

  return (
    <MonthlyStatsContext.Provider value={value}>
      {children}
    </MonthlyStatsContext.Provider>
  );
}

export function useMonthlyStats(): MonthlyStatsValue {
  const value = useContext(MonthlyStatsContext);

  if (!value) {
    throw new Error('useMonthlyStats() must be used in a MonthlyStatsProvider');
  }

  return value;
}
