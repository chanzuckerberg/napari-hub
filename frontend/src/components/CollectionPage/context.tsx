import { createContext, ReactNode, useContext } from 'react';

import { CollectionData } from '@/types/collections';

const CollectionContext = createContext<CollectionData | null>(null);

interface Props {
  collection: CollectionData;
  children: ReactNode;
}

/**
 * Provider for access to collection state from anywhere in collection page.
 */
export function CollectionContextProvider({ collection, children }: Props) {
  return (
    <CollectionContext.Provider value={collection}>
      {children}
    </CollectionContext.Provider>
  );
}

/**
 * Hook for accessing collection data from hub API.
 */
export function useCollection(): CollectionData {
  const collection = useContext(CollectionContext);

  if (!collection) {
    throw new Error(
      'useCollections() may not be used outside of CollectionContextProvider',
    );
  }

  return collection;
}
