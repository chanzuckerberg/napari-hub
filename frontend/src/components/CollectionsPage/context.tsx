import { createContext, ReactNode, useContext } from 'react';

import { CollectionIndexData } from '@/types/collections';

const CollectionsContext = createContext<CollectionIndexData[]>([]);

interface Props {
  collections: CollectionIndexData[];
  children: ReactNode;
}

/**
 * Provider for access to collections state from anywhere in collections home
 * page UI.
 */
export function CollectionsContextProvider({ collections, children }: Props) {
  return (
    <CollectionsContext.Provider value={collections}>
      {children}
    </CollectionsContext.Provider>
  );
}

/**
 * Hook for accessing list of collections from hub API.
 */
export function useCollections(): CollectionIndexData[] {
  return useContext(CollectionsContext);
}
