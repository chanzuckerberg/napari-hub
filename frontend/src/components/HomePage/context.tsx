import { createContext, ReactNode, useContext, useMemo } from 'react';

import { PluginSectionsResponse } from '@/types';

interface HomePageContextValue {
  pluginSections: PluginSectionsResponse;
}

const HomePageContext = createContext<HomePageContextValue | null>(null);

interface Props extends Pick<HomePageContextValue, 'pluginSections'> {
  children: ReactNode;
}

export function HomePageProvider({ children, pluginSections }: Props) {
  const value = useMemo(
    () => ({
      pluginSections,
    }),
    [pluginSections],
  );

  return (
    <HomePageContext.Provider value={value}>
      {children}
    </HomePageContext.Provider>
  );
}

export function useHomePage(): HomePageContextValue {
  const value = useContext(HomePageContext);

  if (!value) {
    throw new Error('useHomePage must be used in a HomePageProvider');
  }

  return value;
}
