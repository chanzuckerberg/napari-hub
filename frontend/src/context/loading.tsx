import { createContext, ReactNode, useContext } from 'react';

const LoadingStateContext = createContext<boolean>(false);

interface Props {
  loading: boolean;
  children: ReactNode;
}

/**
 * Provider for providing global loading state to a component tree.
 */
export function LoadingStateProvider({ children, loading }: Props) {
  return (
    <LoadingStateContext.Provider value={loading}>
      {children}
    </LoadingStateContext.Provider>
  );
}

/**
 * Returns the global loading state.
 *
 * @returns The loading state.
 */
export function useLoadingState(): boolean {
  return useContext(LoadingStateContext);
}
