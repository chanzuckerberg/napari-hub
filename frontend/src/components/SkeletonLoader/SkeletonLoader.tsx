import Skeleton from '@mui/lab/Skeleton';
import { ReactNode } from 'react';

import { useLoadingState } from '@/context/loading';

interface Props {
  /**
   * Class to apply to Skeleton loader.
   */
  className?: string;

  /**
   * Render function that returns the node to render when not loading.
   *
   * https://reactjs.org/docs/render-props.html
   */
  render(): ReactNode;
}

/**
 * Component that renders a skeleton loader when the global loading state is `true`.
 */
export function SkeletonLoader({ className, render }: Props) {
  const isLoading = useLoadingState();
  if (!isLoading) return <>{render()}</>;
  return (
    <Skeleton
      className={className}
      data-testid="skeleton-loader"
      variant="rectangular"
    />
  );
}
