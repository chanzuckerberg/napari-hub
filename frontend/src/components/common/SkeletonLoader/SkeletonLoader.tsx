import { Skeleton } from '@material-ui/lab';
import { useAtom } from 'jotai';
import { ReactNode } from 'react';

import { loadingState } from '@/store/loading';

interface Props {
  /**
   * Class to apply to Skeleton loader.
   */
  className?: string;

  /**
   * Render function that returns the node to render when not loading.
   *
   * https://reactjs.org/docs/render-props.html
   */
  render(): ReactNode;
}

/**
 * Component that renders a skeleton loader when the global loading state is `true`.
 */
export function SkeletonLoader({ className, render }: Props) {
  const [isLoading] = useAtom(loadingState);
  if (!isLoading) return <>{render()}</>;
  return (
    <Skeleton
      className={className}
      data-testid="skeleton-loader"
      variant="rect"
    />
  );
}
