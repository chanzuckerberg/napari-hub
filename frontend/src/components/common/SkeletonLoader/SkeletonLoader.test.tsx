import { render } from '@testing-library/react';

import { LoadingStateProvider } from '@/context/loading';

import { SkeletonLoader } from './SkeletonLoader';

describe('<SkeletonLoader />', () => {
  const text = 'Hello, World!';

  it('should match snapshot', () => {
    let component = render(<SkeletonLoader render={() => <h1>{text}</h1>} />);
    expect(component.asFragment()).toMatchSnapshot();

    component = render(
      <LoadingStateProvider loading={false}>
        <SkeletonLoader render={() => <h1>{text}</h1>} />
      </LoadingStateProvider>,
    );
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should render skeleton when loading', () => {
    const component = render(
      <LoadingStateProvider loading>
        <SkeletonLoader render={() => <h1>{text}</h1>} />
      </LoadingStateProvider>,
    );

    expect(component.getByTestId('skeleton-loader')).toBeDefined();
  });
});
