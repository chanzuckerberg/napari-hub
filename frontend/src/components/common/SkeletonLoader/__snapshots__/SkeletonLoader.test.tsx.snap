// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SkeletonLoader /> should match snapshot 1`] = `
<DocumentFragment>
  <h1>
    Hello, World!
  </h1>
</DocumentFragment>
`;

exports[`<SkeletonLoader /> should match snapshot 2`] = `
<DocumentFragment>
  <h1>
    Hello, World!
  </h1>
</DocumentFragment>
`;
